lib/ts/MemArena.cc

```diff
@@ -0,0 +1,221 @@
+/** @file
+
+    MemArena memory allocator. Chunks of memory are allocated, frozen into generations and
+     thawed away when unused.
+
+    @section license License
+
+    Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#include "MemArena.h"
+#include <ts/ink_memory.h>
+#include <ts/ink_assert.h>
+
+using namespace ts;
+
+inline MemArena::Block::Block(size_t n) : size(n), allocated(0), next(nullptr)
+{
+}
+
+inline char *
+MemArena::Block::data()
+{
+  return reinterpret_cast<char *>(this + 1);
+}
+
+/**
+    Allocates a new internal block of memory. If there are no existing blocks, this becomes the head of the
+     @allocations ll. If there are existing allocations, the new block is inserted in the allocations list.
+     If @back == true, the new block is appended and becomes the @current internal block.
+        @back == false, the new block is pushed into the generation but @current doesn't change.
+  */
+inline MemArena::Block *
+MemArena::newInternalBlock(size_t n, bool back)
+{
+  // Adjust to the nearest power of two. Works for 64 bit values. Allocate Block header and
+  //  actual underlying memory together for locality. ALLOC_HEADER_SIZE to account for malloc/free headers.
+  static constexpr size_t free_space_per_page = DEFAULT_PAGE_SIZE - sizeof(Block) - ALLOC_HEADER_SIZE;
+
+  void *tmp;
+  if (n <= free_space_per_page) { // will fit within one page, just allocate.
+    tmp = ats_malloc(n + sizeof(Block));
+  } else {
+    size_t t = n;
+    t--;
+    t |= t >> 1;
+    t |= t >> 2;
+    t |= t >> 4;
+    t |= t >> 8;
+    t |= t >> 16;
+    t |= t >> 32;
+    t++;
+    n   = t - sizeof(Block) - ALLOC_HEADER_SIZE; // n is the actual amount of memory the block can allocate out.
+    tmp = ats_malloc(t - ALLOC_HEADER_SIZE);
+  }
+
+  std::shared_ptr<Block> block(new (tmp) Block(n)); // placement new
+
+  if (allocations) {
+    arena_size += n;
+    generation_size += n;
+
+    if (back) { // append
+      current->next = block;
```

Comment:
Here, if you did
```
block->next = current;
current = block
```
The chain would be anchored by the most recent block and you would only need that to both find the most recent block and maintain the chain. Then `current` and `allocations` would be the same thing. For a "custom" block for an  allocation, you do
```
if (current) {
  block->next = current->next;
  current->next = block;
} else {
  current = block;
}
```
This leaves `current` the current block. Then you only need `current` and `generation`.

Reply:
Good point. Removed `allocations`. List now grows from head. Thawing just nulls `generation->next` if `generation` exists. Updated docs.